actions/dataset.go

```diff
@@ -12,56 +13,74 @@ import (
 	"github.com/qri-io/qri/base"
 	"github.com/qri-io/qri/p2p"
 	"github.com/qri-io/qri/repo"
+	"github.com/qri-io/qri/repo/profile"
 )
 
 // SaveDataset initializes a dataset from a dataset pointer and data file
 func SaveDataset(node *p2p.QriNode, dsp *dataset.DatasetPod, dryRun, pin bool) (ref repo.DatasetRef, body cafs.File, err error) {
 	var (
-		ds       *dataset.Dataset
-		bodyFile cafs.File
-		secrets  map[string]string
-		// NOTE - struct fields need to be instantiated to make assign set to
-		// new pointer values
-		userSet = &dataset.Dataset{
-			Commit:    &dataset.Commit{},
-			Meta:      &dataset.Meta{},
-			Structure: &dataset.Structure{},
-			Transform: &dataset.Transform{},
-			Viz:       &dataset.Viz{},
-		}
+		ds                       *dataset.Dataset
+		prevPath                 string
+		bodyFile, changeBodyFile cafs.File
+		secrets                  map[string]string
+		pro                      *profile.Profile
+		changes                  = &dataset.Dataset{}
+		r                        = node.Repo
 	)
 
+	// set ds to dataset head, or empty dataset if no history
+	ds, bodyFile, prevPath, err = base.PrepareDatasetSave(r, dsp.Peername, dsp.Name)
```

Comment:
Not totally clear that ds is the fully hydrated previous dataset, with just the PreviousPath field updated. It makes it a bit hard to understand the actions we take going forward without knowing this.